File: modules/audio-focus/src/AudioFocusModule.ts
import { requireNativeModule } from "expo";

const AudioFocusModule = requireNativeModule("AudioFocus");

/**
 * Requests Android audio focus with AUDIOFOCUS_GAIN.
 * This causes other apps (Spotify, podcasts, YouTube Music) to pause.
 * No-op on iOS (handled by expo-audio's setAudioModeAsync).
 */
export async function requestAudioFocus(): Promise<void> {
  return AudioFocusModule.requestAudioFocus();
}

/**
 * Abandons Android audio focus so other apps can resume.
 * No-op on iOS.
 */
export async function abandonAudioFocus(): Promise<void> {
  return AudioFocusModule.abandonAudioFocus();
}

File: modules/audio-focus/ios/AudioFocusModule.swift
import ExpoModulesCore
import AVFoundation

public class AudioFocusModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AudioFocus")

    // Activate the AVAudioSession so other apps (Spotify, podcasts) are interrupted.
    // setAudioModeAsync has already set the category/options via expo-audio;
    // we just flip the session active here.
    AsyncFunction("requestAudioFocus") {
      try AVAudioSession.sharedInstance().setActive(true)
    }

    // Deactivate with .notifyOthersOnDeactivation — the iOS equivalent of
    // abandonAudioFocus on Android. This is the signal that tells Spotify /
    // podcasts / music apps that they can resume playback.
    AsyncFunction("abandonAudioFocus") {
      try AVAudioSession.sharedInstance().setActive(
        false,
        options: .notifyOthersOnDeactivation
      )
    }
  }
}
